core/src/main/java/com/datastax/oss/driver/api/core/session/SessionBuilder.java

```diff
@@ -633,73 +637,103 @@ public SessionT build() {
     return CompletableFutures.getUninterruptibly(buildAsync());
   }
 
+  /**
+   * Creates an uninitialized session with the options set by this builder.
+   *
+   * <p>The returned session will be disconnected; it must be manually {@linkplain
+   * LazyCqlSession#init() initialized} before it is used to perform queries.
+   *
+   * @throws UnsupportedOperationException it this builder does not support creating uninitialized
+   *     sessions.
+   */
+  @NonNull
+  public SessionT buildLazy() {
+    LazyCqlSession defaultSession = buildDefaultSessionLazy();
+    return wrapLazy(defaultSession);
+  }
+
   protected abstract SessionT wrap(@NonNull CqlSession defaultSession);
 
+  protected SessionT wrapLazy(@NonNull LazyCqlSession defaultSession) {
+    throw new UnsupportedOperationException("Building uninitialized sessions not supported");
+  }
+
   @NonNull
   protected final CompletionStage<CqlSession> buildDefaultSessionAsync() {
     try {
+      LazyCqlSession defaultSession = buildDefaultSessionLazy();
+      return defaultSession.init().thenApply(v -> defaultSession);
+    } catch (Throwable t) {
+      // We construct the session synchronously (until the init() call), but async clients expect a
+      // failed future if anything goes wrong. So wrap any error from that synchronous part.
+      return CompletableFutures.failedFuture(t);
+    }
+  }
 
-      ProgrammaticArguments programmaticArguments = programmaticArgumentsBuilder.build();
+  @NonNull
+  protected LazyCqlSession buildDefaultSessionLazy() {
```

Comment:
`buildDefaultSessionAsync` is final, but I don't see why it should be; and making this method final would make revapi angry.